### PR TITLE
Network Providers

### PR DESCRIPTION
**Pull request by popenta**
Added network providers for both API and Proxy similar to sdk-erdjs-network-providers.
Also implemented specialized methods to work with api & proxy.

### CODE REVIEW

File: erdpy_network/tokens.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+from erdpy_core import Address
+from typing import Any, List, Dict
+from erdpy_network.primitives import Nonce
+
+
+class FungibleTokenOfAccountOnNetwork:
+    def __init__(self):
+        self.identifier: str = ''
+        self.balance: int = 0
+        self.raw_response: Any = {}
```
Comment (andreibancioiu) [inline erdpy_network/tokens.py:11]: I see, we also keep the original response here. Seems all right. Let's think if we should do this for all the other objects, as well - in the future.

File: erdpy_network/transactions.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+class TransactionOnNetwork:
+    def __init__(self):
+        self.is_completed: bool = False
+        self.hash: str = ''
+        self.type: str = ''
+        self.nonce: int = 0
+        self.round: int = 0
+        self.epoch: int = 0
+        self.value: str = ''
```
Comment (andreibancioiu) [inline erdpy_network/transactions.py:21]: Can be `int`.

File: erdpy_network/utils.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+def decimal_to_padded_hex(i):
+    return "0" + f"{i:x}" if len(f"{i:x}") % 2 else f"{i:x}"
```
Comment (andreibancioiu) [inline erdpy_network/utils.py:2]:
```
as_hex = f"{i:x}"
as_hex_padded = "0" + as_hex if ...
```

(that is, suggestion to extract a temporary variable `as_hex`)

File: erdpy_network/accounts.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    @staticmethod
+    def from_http_response(payload: Dict[str, Any]) -> 'AccountOnNetwork':
+        result = AccountOnNetwork()
+
+        address = payload.get('address', '')
+        result.address = Address.from_bech32(address) if address else Address.zero()
+
+        result.nonce = payload.get('nonce', 0)
+        result.balance = str(payload.get('balance', ''))
+        result.code = bytes(payload.get('code', '').encode())
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/accounts.py:23]: Indeed it's returned as hex. Changed!

File: erdpy_network/api_network_provider.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+    def __build_pagination_params(self, pagination: IPagination):
+        return f'from={pagination.get_start()}&size={pagination.get_size()}'
+
+    def do_get_generic(self, resource_url: str):
+        response = self.do_get(f'{self.url}/{resource_url}')
+        return response
+
+    def do_post_generic(self, resource_url: str, payload: Any):
+        response = self.do_post(resource_url, payload)
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/api_network_provider.py:147]: Done!

File: erdpy_network/api_network_provider.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        except requests.ConnectionError as err:
+            raise GenericError(url, err)
+        except Exception as err:
+            raise GenericError(url, err)
+
+    def _get_data(self, parsed: Union[Dict, List], url: str):
+        if isinstance(parsed, List):
+            return parsed
+        else:
+            err = parsed.get("error", None)
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/api_network_provider.py:184]: Added test.

File: erdpy_network/api_network_provider_test.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
+from erdpy_core import Address
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/api_network_provider_test.py:1]: Added pytest as development dependency in the readme.

File: erdpy_network/api_network_provider_test.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+        assert result.address.bech32() == 'erd1testnlersh4z0wsv8kjx39me4rmnvjkwu8dsaea7ukdvvc9z396qykv7z7'
+        assert result.username == ''
+
+    def test_get_fungible_token_of_account(self):
+        address = Address.from_bech32('erd1testnlersh4z0wsv8kjx39me4rmnvjkwu8dsaea7ukdvvc9z396qykv7z7')
+        result = self.api.get_fungible_token_of_account(address, 'ABC-10df96')
+
+        assert result.identifier == 'ABC-10df96'
+        assert result.balance == 50
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/api_network_provider_test.py:39]: I agree. I'm thinking about deleting it.

File: erdpy_network/api_network_provider_test.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+    def test_get_nonfungible_token_of_account(self):
+        address = Address.from_bech32('erd1testnlersh4z0wsv8kjx39me4rmnvjkwu8dsaea7ukdvvc9z396qykv7z7')
+        result = self.api.get_nonfungible_token_of_account(address, 'ASDASD-510041', 2)
+
+        assert result.balance == 0
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/api_network_provider_test.py:45]: The api does not return a balance attribute, so it initialises as default with 0.

File: erdpy_network/config.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
+from erdpy_network.interface import IPagination
+from erdpy_core import Address
+
+ESDT_CONTRACT_ADDRESS = Address.from_bech32("erd1qqqqqqqqqqqqqqqpqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqzllls8a5w6u")
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/config.py:4]: Moved it in a file called `constants.py` as it's the same in `erdjs-network-providers`.

File: erdpy_network/contract_query_response.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+class ContractQueryResponse:
+    def __init__(self):
+        self.return_data: List[str] = []
+        self.return_code: str = ''
+        self.return_message: str = ''
+
+    @staticmethod
+    def from_http_response(payload: Dict[str, Any]) -> 'ContractQueryResponse':
+        return_data = payload.get('returnData', '') or payload.get('ReturnData', '')
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/contract_query_response.py:13]: Replaced empty string with empty list.

File: erdpy_network/contract_query_response.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+        result = ContractQueryResponse()
+        result.return_data = return_data
+        result.return_code = return_code
+        result.return_message = return_message
+
+        return result
+
+    def get_return_data_parts(self) -> List:
+        return list(map(lambda x: base64.b64decode(x), self.return_data))
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/contract_query_response.py:26]: Replaced it with a list comprehension.

File: erdpy_network/contract_results.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+
+class ContractResults:
+    def __init__(self, items: List['ContractResultItem']):
+        self.items = items
+        self.items.sort(key=lambda x: x.nonce, reverse=False)
+
+    @staticmethod
+    def from_api_http_response(results: List[Any]) -> 'ContractResults':
+        items = list(map(ContractResultItem.from_api_http_response, results))
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/contract_results.py:15]: Replaced it with a list comprehension.

File: erdpy_network/contract_results.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        return item
+
+    @staticmethod
+    def from_proxy_http_response(response: Any) -> 'ContractResultItem':
+        item = ContractResultItem.from_http_response(response)
+
+        return item
+
+    @staticmethod
+    def from_http_response(response: Dict[str, Any]) -> 'ContractResultItem':
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/contract_results.py:56]: Done!

File: erdpy_network/contract_results.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+        receiver = response.get('receiver', '')
+        item.receiver = Address.from_bech32(receiver) if receiver else Address.zero()
+
+        item.previous_hash = response.get('prevTxHash')
+        item.original_hash = response.get('originalTxHash')
+        item.gas_limit = response.get('gasLimit', 0)
+        item.gas_price = response.get('gasPrice', 0)
+        item.data = response.get('data', '')
+        item.call_type = response.get('callType')
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/contract_results.py:74]: Added 0 as default.

File: erdpy_network/interface.py [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+class ITransaction(ISerializable, Protocol):
+    def to_dictionary(self) -> Dict[str, Any]:
+        return {}
+
+    def get_hash(self) -> str:
+        return ""
+
+
+class ISimulateResponse(ISerializable):
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/interface.py:22]: They are not used. Removed!

File: erdpy_network/primitives.py [hunk 17]
```diff
@@ -0,0 +1,9 @@
+from erdpy_network.utils import decimal_to_padded_hex
+
+
+class Nonce:
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/primitives.py:4]: Deleted the class, now only the `decimal_to_padded_hex` function used

File: erdpy_network/utils.py (hunk 3 above)
Comment (popenta) [reply erdpy_network/utils.py:2]: Done

File: erdpy_network/api_network_provider.py [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+
+    def get_nonfungible_token_of_account(self, address: IAddress, collection: str, nonce: int):
+        nonce_as_hex = Nonce(nonce).hex()
+        url = f'accounts/{address.bech32()}/nfts/{collection}-{nonce_as_hex}'
+        response = self.do_get_generic(url)
+        result = NonFungibleTokenOfAccountOnNetwork.from_api_http_response(response)
+
+        return result
+
+    def get_mex_pairs(self, pagination: IPagination = None):
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/api_network_provider.py:85]: Deleted!

File: erdpy_network/proxy_network_provider.py [hunk 19]
```diff
@@ -27,7 +27,7 @@ def get_network_config(self) -> NetworkConfig:
         return network_config
 
     def get_network_status(self) -> NetworkStatus:
-        response = self.do_get_generic('network/status/4294967295')
+        response = self.do_get_generic(f'network/status/{str(METACHAIN_ID)}')
```
Comment (andreibancioiu) [inline erdpy_network/proxy_network_provider.py:30]: `str()` is not necessary here.

**Review by andreibancioiu — DISMISSED** [review]
:+1:

File: erdpy_network/proxy_network_provider.py (hunk 19 above)
Comment (popenta) [reply erdpy_network/proxy_network_provider.py:30]: Removed it.

File: erdpy_network/api_network_provider_test.py [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+        assert result.active_validators > 0
+        assert result.total_staked > 0
+
+    def test_get_general_statistics(self):
+        result = self.api.get_network_general_statistics()
+
+        assert result.shards == 3
+        assert result.rounds_per_epoch == 1200
+        assert result.refresh_rate == 6000
+        assert result.epoch >= 2864
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/api_network_provider_test.py:24]: I was aware of that. Maybe we'll modify them in the future.

File: erdpy_network/proxy_network_provider.py [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+        esdts = [items[key] for key in items.keys() if items[key].get('nonce', '') == '']
+        tokens = map(FungibleTokenOfAccountOnNetwork.from_http_response, esdts)
+
+        return list(tokens)
+
+    def get_nonfungible_tokens_of_account(self, address: IAddress) -> List[NonFungibleTokenOfAccountOnNetwork]:
+        url = f'address/{address.bech32()}/esdt'
+        response = self.do_get_generic(url)
+        items = response.get('esdts')
+        nfts = [items[key] for key in items.keys() if items[key].get('nonce', -1) >= 0]
```
[the remark this answers is not in the document]
Comment (popenta) [reply erdpy_network/proxy_network_provider.py]: Changed condition to be greater than 0.

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**